package/src/ssr/NextSSRApolloClient.tsx

```diff
@@ -0,0 +1,169 @@
+import {
+  ApolloBackgroundQueryTransport,
+  ApolloResultCache,
+} from "./ApolloRehydrateSymbols";
+
+function getQueryManager<TCacheShape>(
+  client: ApolloClient<unknown>
+): QueryManager<TCacheShape> {
+  return client["queryManager"];
+}
+
+export class NextSSRApolloClient<
+  TCacheShape
+> extends ApolloClient<TCacheShape> {
+  private rehydrationContext: Pick<
+    RehydrationContextValue,
+    "incomingBackgroundQueries"
+  > = {
+    incomingBackgroundQueries: [],
+  };
+
+  constructor(options: ApolloClientOptions<TCacheShape>) {
+    super(options);
+
+    this.registerWindowHook();
+  }
+
+  private resolveFakeQueries = new Map<
+    string,
+    [(result: FetchResult) => void, (reason: any) => void]
+  >();
+
+  private identifyUniqueQuery(options: {
+    query: DocumentNode;
+    variables?: unknown;
+  }) {
+    const transformedDocument = this.documentTransform.transformDocument(
+      options.query
+    );
+    const queryManager = getQueryManager<TCacheShape>(this);
+    // Calling `transformDocument` will add __typename but won't remove client
+    // directives, so we need to get the `serverQuery`.
+    const { serverQuery } = queryManager.getDocumentInfo(transformedDocument);
+
+    const canonicalVariables = canonicalStringify(options.variables);
+
+    const cacheKey = [serverQuery, canonicalVariables].toString();
+
+    return { query: serverQuery, cacheKey, varJson: canonicalVariables };
+  }
+
+  private registerWindowHook() {
+    if (typeof window !== "undefined") {
+      if (Array.isArray(window[ApolloBackgroundQueryTransport] || [])) {
+        registerLateInitializingQueue(
+          ApolloBackgroundQueryTransport,
+          (options) => {
+            const { query, varJson, cacheKey } =
+              this.identifyUniqueQuery(options);
+
+            if (!query) return;
+            const printedServerQuery = print(query);
+            const queryManager = getQueryManager<TCacheShape>(this);
+
+            const byVariables =
+              queryManager["inFlightLinkObservables"].get(printedServerQuery) ||
+              new Map();
+
+            queryManager["inFlightLinkObservables"].set(
+              printedServerQuery,
+              byVariables
+            );
+
+            if (!byVariables.has(varJson)) {
+              const promise = new Promise<FetchResult>((resolve, reject) => {
+                this.resolveFakeQueries.set(cacheKey, [resolve, reject]);
+              });
+
+              byVariables.set(
+                varJson,
+                new Observable<FetchResult>((observer) => {
+                  promise
+                    .then((result) => {
+                      observer.next(result);
+                      observer.complete();
+                    })
+                    .catch((err) => {
+                      observer.error(err);
+                    }).finally(() => {
+                      this.resolveFakeQueries.delete(cacheKey);
+                    });
```

Comment:
```suggestion
                    }).finally(() => {
                      this.resolveFakeQueries.delete(cacheKey);
                      cleanupCancelFn();
                    });
```

For comparison:
https://github.com/apollographql/apollo-client/blob/d64a9656ee745349bb4958a743065f2a91c2d0cc/src/core/QueryManager.ts#L1278

Reply:
Resolved in [6462b2a](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/6462b2a8c2f1dfb24a66a34f77944cb1a9436546)